metrics/internal/collectors/storage-auto-scaler.go

```diff
@@ -0,0 +1,174 @@
+			storageAutoScaler.Namespace,
+			storageAutoScaler.Spec.StorageCluster.Name,
+			storageAutoScaler.Spec.DeviceClass,
+		)
+	}
+}
+
+func (s *StorageAutoScalerCollector) collectStorageAutoScalerPhaseStatuses(ch chan<- prometheus.Metric, storageAutoScalers []*v1.StorageAutoScaler, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for _, storageAutoScaler := range storageAutoScalers {
+		var phaseStatus = NotStartedPhaseStatus
+		switch storageAutoScaler.Status.Phase {
+		case v1.StorageAutoScalerPhaseFailed:
+			phaseStatus = FailedPhaseStatus
+		case v1.StorageAutoScalerPhaseSucceeded:
+			phaseStatus = succeededPhaseValidation(storageAutoScaler.Status.LastExpansion.CompletionTime)
+		case v1.StorageAutoScalerPhaseInProgress:
+			phaseStatus = InProgressPhaseStatus
+		default:
+			phaseStatus = NotStartedPhaseStatus
+		}
+		// send metric on any phase other than 'Reset' phase
+		if phaseStatus != ResetPhaseStatus {
+			ch <- prometheus.MustNewConstMetric(
+				s.AutoScalerPhaseStatus,
+				prometheus.GaugeValue,
+				float64(phaseStatus),
+				storageAutoScaler.Name,
+				storageAutoScaler.Namespace,
+			)
+		}
+	}
+}
+
+func succeededPhaseValidation(completionTime metav1.Time) PhaseStatus {
+	currentTime := metav1.Now().Time
+	var zeroValuedTime metav1.Time
+	// if completion time is not set
+	// OR
+	// success happened within a 10mins time window
+	// we return success
+	if completionTime.Sub(zeroValuedTime.Time) == 0 ||
+		currentTime.Sub(completionTime.Time) < 10*time.Minute {
+		return SucceededPhaseStatus
+	}
+	// after 10 mins, we will reset the success phase
+	return ResetPhaseStatus
+}
+
+func (s *StorageAutoScalerCollector) collectStorageCapacityReachedStatuses(ch chan<- prometheus.Metric, storageAutoScalers []*v1.StorageAutoScaler, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for _, storageAutoScaler := range storageAutoScalers {
+		var capacityReached = 0
+		if storageAutoScaler.Status.StorageCapacityLimitReached != nil && *storageAutoScaler.Status.StorageCapacityLimitReached {
+			capacityReached = 1
+		}
+		ch <- prometheus.MustNewConstMetric(
+			s.StorageCapacityReached,
+			prometheus.GaugeValue,
+			float64(capacityReached),
+			storageAutoScaler.Name,
+			storageAutoScaler.Namespace,
+		)
+	}
+}
+
+func NewStorageAutoScalerCollector(opts *options.Options) *StorageAutoScalerCollector {
+	c, err := GetOcsV1Client(opts)
+	if err != nil {
+		klog.Errorf("Unable to get client: %v", err)
+		return nil
+	}
+	lw := cache.NewListWatchFromClient(c, "storageautoscalers", searchInNamespace(opts), fields.Everything())
+	sharedIndexInformer := cache.NewSharedIndexInformer(lw, &v1.StorageAutoScaler{}, 0, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+	return &StorageAutoScalerCollector{
+		StorageAutoScalerMetadata: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "storageautoscaler", "metadata"),
+			"Contains StorageAutoScaler details",
+			[]string{"name", "namespace", "storagecluster", "device_class"},
```

Comment:
Can these labels be added to each metrics to avoid generating this metadata metrics?

Reply:
Done, please check